src/bt.rs

```diff
@@ -0,0 +1,487 @@
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<P, V: ParameterizedDecode<P>> ParameterizedDecode<P> for Node<V> {
+    fn decode_with_param(param: &P, bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let mut root = None;
+        let mut stack = vec![&mut root];
+
+        // Decode nodes in a pre-order traversal.
+        while let Some(elem) = stack.pop() {
+            match CodecMarker::decode(bytes)? {
+                CodecMarker::Leaf => (),
+                CodecMarker::Inner => {
+                    *elem = Some(Box::new(Node::new(V::decode_with_param(param, bytes)?)));
+                    let node = elem.as_mut().unwrap();
+                    stack.push(&mut node.right);
+                    stack.push(&mut node.left);
+                }
+            };
+        }
+
+        match root {
+            Some(node) => Ok(*node),
+            None => Err(CodecError::UnexpectedValue),
+        }
+    }
+}
+
+#[cfg(feature = "test-util")]
+impl<V: core::fmt::Display> core::fmt::Display for Node<V> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> std::fmt::Result {
+        struct Item<'a, V> {
+            name: String,
+            level: usize,
+            node: Option<&'a Node<V>>,
+        }
+
+        let mut stack = vec![Item {
+            name: String::default(),
+            level: 0,
+            node: Some(self),
+        }];
+
+        while let Some(Item { name, level, node }) = stack.pop() {
+            if let Some(Node { value, left, right }) = node {
+                let prefix = "  ".repeat(level);
+                writeln!(f, "{name}\n{prefix}node: {value}")?;
+
+                stack.push(Item {
+                    name: format!("{prefix}right:"),
+                    level: level + 1,
+                    node: right.as_deref(),
+                });
+
+                stack.push(Item {
+                    name: format!("{prefix}left:"),
+                    level: level + 1,
+                    node: left.as_deref(),
+                });
+            } else {
+                writeln!(f, "{name} <None>")?
+            };
+        }
+
+        Ok(())
+    }
+}
+
+impl<V: Encode> Encode for BinaryTree<V> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        match &self.root {
+            None => CodecMarker::Leaf.encode(bytes),
+            Some(node) => node.encode(bytes),
+        }
+    }
+}
+
+impl<P, V: ParameterizedDecode<P>> ParameterizedDecode<P> for BinaryTree<V> {
+    fn decode_with_param(param: &P, bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        match CodecMarker::decode(bytes)? {
+            CodecMarker::Leaf => Ok(Self::default()),
+            CodecMarker::Inner => {
+                bytes.seek(SeekFrom::Start(0))?;
+                let root = Some(Box::new(Node::decode_with_param(param, bytes)?));
+                Ok(Self { root })
+            }
+        }
+    }
+}
```

Comment:
The asymmetry between encoding and decoding, and the seek, make me a bit uneasy. This is ultimately because the encoding of a Node must start with a 1 byte, so BinaryTree uses that byte to signal a non-empty root node, and writes/reads its own 0 byte otherwise for an empty node. But, in the 1 case, that first byte has to be read by the `BinaryTree` impl once, then read again by the `Node` impl after rewinding.

I'd like to get rid of the seeking to make future maintenance easier, particularly in case we make further big changes to the codec traits. Moreover, if a `BinaryTree<V>` were included as a field in a larger message, seeking to the start of the buffer would be incorrect -- we'd instead need to save `bytes.stream_position()` before reading the `CodecMarker`, and rewind by seeking to that position.

I suggest either eliminating the codec impls on `Node<V>`, combining them with these, and then initializing the `BinaryTree` impl's encoding stack with `vec![self.root.as_ref()]`, or moving common codec logic into private functions that are used by impls for both `Node<V>` and `BinaryTree<V>`. If we keep codec impls on `Node<V>`, I'd suggest redefining the serialization format such that a node doesn't start with a `CodecMarker`, but a `CodecMarker` still appears before each child slot. This might require changing what enum is stored in the stack, but maybe there's a more clever way to handle this.

Reply:
it wasn't so difficult to remove the seek call after all while keeping serialization of `Node<V>`.